### ISSUE TITLE
Repeated return in ArrayPool

### ISSUE THREAD

**Issue by minghao-guo** [dotnet/runtime#110582]
### Description

In ArrayPool, if extra Return() is called, the next rent array will be immediately returned.

### Reproduction Steps

```
using System;
using System.Buffers;

var arr = ArrayPool<int>.Shared.Rent(1000);
arr[0] = 10;
ArrayPool<int>.Shared.Return(arr);
ArrayPool<int>.Shared.Return(arr);
ArrayPool<int>.Shared.Return(arr);

var arr1 = ArrayPool<int>.Shared.Rent(1000);
var arr2 = ArrayPool<int>.Shared.Rent(1000);
var arr3 = ArrayPool<int>.Shared.Rent(1000);
var arr4 = ArrayPool<int>.Shared.Rent(1000);
arr1[0] = 1;
arr2[0] = 2;
arr3[0] = 3;
arr4[0] = 4;

Console.WriteLine($"{arr1[0]},{arr2[0]},{arr3[0]},{arr4[0]}");
```

### Expected behavior

arr1, arr2, arr3, arr4 are rent sequentially. We expect they are 4 different arrays so the result of Console.WriteLine should be `1,2,3,4`

### Actual behavior

The result of Console.WriteLine is `3,3,3,4`

Since there are 2 extra Return calls, arr1 and arr2 are immediately returned after rent so next Rent call will reuse the same array. arr1, arr2, arr3 all point to same array.

### Known Workarounds

Problem does not exist if create a new ArrayPool instance instead of using the default shared pool implementation

```
using System;
using System.Buffers;

var pool = ArrayPool<int>.Create();

var arr = pool.Rent(1000);
arr[0] = 10;
pool.Return(arr);
pool.Return(arr);
pool.Return(arr);

var arr1 = pool.Rent(1000);
var arr2 = pool.Rent(1000);
var arr3 = pool.Rent(1000);
var arr4 = pool.Rent(1000);
arr1[0] = 1;
arr2[0] = 2;
arr3[0] = 3;
arr4[0] = 4;

Console.WriteLine($"{arr1[0]},{arr2[0]},{arr3[0]},{arr4[0]}");
```

**Comment by huoyaoyuan** [comment]
> Since there are 2 extra Return calls, arr1 and arr2 are immediately returned after rent so next Rent call will reuse the same array. arr1, arr2, arr3 all point to same array.

This is not correct explanation of the observation. Shared `ArrayPool` implementation starts with no element, when `Return` is called, the returned array will be registered for subsequent renting. Thus, there will be duplicated rentable record of `arr`.

Duplicated return is strictly misuse of array pool and will cause undefined behavior. With #109695, you will be able to detect such misuse.

**Comment by minghao-guo** [comment]
> This is not correct explanation of the observation. Shared `ArrayPool` implementation starts with no element, when `Return` is called, the returned array will be registered for subsequent renting. Thus, there will be duplicated rentable record of `arr`.


This behavior is not observed with create a new instance using `var pool = ArrayPool<int>.Create();`  Is it safer to do this or still undefined?

**Comment by Clockwork-Muse** [comment]
Note that this sort of thing is effectively a [double-free vulnerability](https://cwe.mitre.org/data/definitions/415.html), so the proper response is to _not do that_ anyways.

**Comment by huoyaoyuan** [comment]
Closing as answered.
